Firestore/core/src/local/leveldb_index_manager.cc

```diff
@@ -380,7 +698,236 @@ void LevelDbIndexManager::UpdateCollectionGroup(
+      util::ComparisonResult cmp = existing_iter->CompareTo(*new_iter);
+      if (cmp == util::ComparisonResult::Ascending) {
+        // The element was removed if the next element in our ordered
+        // walkthrough is only in `existing`.
+        removed = true;
+      } else if (cmp == util::ComparisonResult::Descending) {
+        // The element was added if the next element in our ordered
+        // walkthrough is only in `new_entries`.
+        added = true;
+      }
+    } else if (existing_iter != existing.cend()) {
+      removed = true;
+    } else {
+      added = true;
+    }
+
+    if (added) {
+      on_add(*new_iter);
+      new_iter++;
+    } else if (removed) {
+      on_remove(*existing_iter);
+      existing_iter++;
+    } else {
+      if (existing_iter != existing.cend()) {
+        existing_iter++;
+      }
+      if (new_iter != new_entries.cend()) {
+        new_iter++;
+      }
+    }
+  }
+}
+
+void LevelDbIndexManager::UpdateEntries(
+    const model::Document& document,
+    const FieldIndex& index,
+    const std::set<IndexEntry>& existing_entries,
+    const std::set<IndexEntry>& new_entries) {
+  DiffSets(
+      existing_entries, new_entries,
+      [this, document, index](const IndexEntry& entry) {
+        this->AddIndexEntry(document, index, entry);
+      },
+      [this, document, index](const IndexEntry& entry) {
+        this->DeleteIndexEntry(document, index, entry);
+      });
+}
+
+void LevelDbIndexManager::AddIndexEntry(const model::Document& document,
+                                        const FieldIndex& index,
+                                        const IndexEntry& entry) {
+  std::string document_key = document->key().path().CanonicalString();
+  auto entry_key = LevelDbIndexEntryKey::Key(
+      entry.index_id(), uid_, entry.array_value(), entry.directional_value(),
+      EncodedDirectionalKey(index, document_key), document_key);
+  db_->current_transaction()->Put(entry_key, "");
+
+  auto document_key_index_prefix =
+      LevelDbIndexEntryDocumentKeyIndexKey::KeyPrefix(entry.index_id(), uid_,
+                                                      document_key);
+  std::string raw_key;
+  auto iter = db_->current_transaction()->NewIterator();
+  for (iter->Seek(document_key_index_prefix); iter->Valid(); iter->Next()) {
+    if (absl::StartsWith(iter->key(), document_key_index_prefix)) {
+      raw_key = iter->key();
+    } else {
+      break;
+    }
+  }
+
+  LevelDbIndexEntryDocumentKeyIndexKey document_key_index_key(
+      entry.index_id(), uid_, document_key, 0);
+  if (!raw_key.empty()) {
+    bool decoded = document_key_index_key.Decode(raw_key);
+    HARD_ASSERT(decoded,
+                "LevelDbIndexEntryDocumentKeyIndexKey cannot be decoded from "
+                "document key index table.");
+    document_key_index_key.IncreaseSeqNumber();
+  }
+
+  db_->current_transaction()->Put(document_key_index_key.Key(), entry_key);
```

Comment:
There is probably no better way, but we are storing the index value twice just so that we can delete it later. Just something I am 99% sure you are aware of, but I want to make sure that you are 100% aware.

Reply:
The indexed values are saved twice in leveldb, one in the index entry table and one in `LevelDbIndexEntryDocumentKeyIndexKey` as an index table.